pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -820,11 +837,15 @@ func (nc *Controller) tryUpdateNodeHealth(node *v1.Node) (time.Duration, v1.Node
 			LastTransitionTime: node.CreationTimestamp,
 		}
 		gracePeriod = nc.nodeStartupGracePeriod
-		nc.nodeHealthMap[node.Name] = &nodeHealthData{
-			status:                   &node.Status,
-			probeTimestamp:           node.CreationTimestamp,
-			readyTransitionTimestamp: node.CreationTimestamp,
-		}
+		if _, found := nc.nodeHealthMap[node.Name]; found {
```

Comment:
When do we expect the node to be in the `nodeHealthMap` but does not have the ready condition set?

Reply:
Another question is whether it's possible that `nc.nodeHealthMap[node.Name]` is set to nil? 

It's not obvious to me that when setting the value in line 929 below `nc.nodeHealthMap[node.Name] = savedNodeHealth`, the `savedNodeHealth` will always be non-nil...

Reply:
Thanks for explaining. I think the code here is not very easy to follow (not this PR's fault). 

One case that seems a bit strange to me is that even if the node doesn't have status, as long as the it continues renewing the lease, the controller will take no action. I think this is by design, but still feels strange :\

Reply:
In terms of heartbeat, I think renewing the lease satisfy the need of making sure node is alive.

@wojtek-t, WDYT?

Reply:
> In terms of heartbeat, I think renewing the lease satisfy the need of making sure node is alive.

I think it does.

Though I kind of agree with @yujuhong that it's a bit strange (but depending how exactly we implement stuff in kubelet (things like retries etc.), I think in the desired implementation it should never happen in practice).

Reply:
According to [KEP-0009](https://github.com/kubernetes/community/blob/master/keps/sig-node/0009-node-heartbeat.md), there will be NodeStatus update at least once per `node-status-update-period` (a new parameter to be added). So the situation where a node does not have status, but keeps renewing lease depends on the value of `node-status-update-period`.

Reply:
Well, there is already `nodeStatusUpdateFrequency`, which controls how often node updates status for now. It will become the one that controls how often node **computes** status later. I would suggest using `nodeStatusReportPeriod` instead of `nodeStatusUpdatePeriod`, as `update` is an overloading term. (Or `nodeStatusReportFrequency`, to be consistent with the update frequency naming convension)

Reply:
> So the situation where a node does not have status, but keeps renewing lease depends on the value of node-status-update-period.

So what we have been thinking is to switch:
- set lease refresh period (however we call it) to how frequently we update node status now
- decrease frequency of node status updates to something like 1 minute
- but still compute node statuses with the same frequency (equal of lease refresh period)

So if we make leases and node-statuses independent, then yes - there may be small period when there is no node status (this can pretty much happen only after node registration though, because after that it should always be there).
But if we would make them somehow correlated (I'm not saying we should do that - I'm saying it's theoretically possible), then it would only happen on races.

But yeah - I agree the latter probably doesn't make much sense.

Reply:
I was more interested in the pathological case where kubelet could not post the status ever, while it could still sending heartbeats. I think this theoretically could happen if we decouple heartbeat and status updates in kubelet (which is probably the right thing to do to not add more complexity and dependency). I don't think we need to deal with the pathological case now, but in case we want to put a safeguard in the future, let's settle on leaving a comment saying something like `If kubelet never posted the node status, but continues renewing the heartbeat leases, the node controller will assume the node is healthy and take no action`.

Reply:
+1

Reply:
Done. Added the comments at line 923 now. PTAL